File: acgs2-core/sdk/python/acgs2_sdk/services/api_gateway.py
"""
ACGS-2 API Gateway Service
Constitutional Hash: cdd01ef066bc6cf2
"""

from typing import TYPE_CHECKING, Any, Dict, List

if TYPE_CHECKING:
    from acgs2_sdk.client import ACGS2Client


class APIGatewayService:
    """Service for interacting with the API Gateway endpoints."""

    def __init__(self, client: "ACGS2Client") -> None:
        """Initialize the API gateway service.

        Args:
            client: ACGS2 client instance
        """
        self._client = client

    async def health_check(self) -> Dict[str, Any]:
        """Check API gateway health.

        Returns:
            Health check response
        """
        return await self._client.get("/health")

    async def submit_feedback(
        self,
        user_id: str,
        category: str,
        rating: int,
        title: str,
        description: str = "",
        metadata: Dict[str, Any] = None,
    ) -> Dict[str, Any]:
        """Submit user feedback.

        Args:
            user_id: User identifier
            category: Feedback category (bug, feature, general)
            rating: Rating from 1-5
            title: Feedback title
            description: Feedback description
            metadata: Additional metadata

        Returns:
            Feedback submission response
        """
        data = {
            "user_id": user_id,
            "category": category,
            "rating": rating,
            "title": title,
            "description": description,
        }

        if metadata:
            data["metadata"] = metadata

        return await self._client.post("/feedback", json=data)

    async def get_feedback_stats(self) -> Dict[str, Any]:
        """Get feedback statistics.

        Returns:
            Feedback statistics
        """
        return await self._client.get("/feedback/stats")

    async def list_services(self) -> Dict[str, Any]:
        """List available services.

        Returns:
            Services information
        """
        return await self._client.get("/services")

File: src/core/sdk/typescript/src/services/api-gateway.ts
/**
 * ACGS-2 API Gateway Service
 * Constitutional Hash: cdd01ef066bc6cf2
 */

import { ACGS2Client } from '../client';
import { CONSTITUTIONAL_HASH } from '../types';
import { ValidationError } from '../utils';

// =============================================================================
// Types
// =============================================================================

export interface HealthCheckResponse {
  status: string;
  version?: string;
  timestamp: string;
  constitutionalHash: string;
}

export interface FeedbackRequest {
  userId: string;
  category: 'bug' | 'feature' | 'general' | 'performance' | 'security';
  rating: number; // 1-5
  title: string;
  description?: string;
  metadata?: Record<string, any>;
}

export interface FeedbackResponse {
  id: string;
  status: 'received' | 'processing' | 'completed';
  timestamp: string;
}

export interface FeedbackStats {
  totalFeedback: number;
  averageRating: number;
  categoryBreakdown: Record<string, number>;
  recentFeedback: FeedbackSummary[];
}

export interface FeedbackSummary {
  id: string;
  userId: string;
  category: string;
  rating: number;
  title: string;
  timestamp: string;
}

export interface ServiceInfo {
  name: string;
  version: string;
  status: 'healthy' | 'degraded' | 'unhealthy';
  endpoints: string[];
  description?: string;
}

export interface ServicesResponse {
  services: ServiceInfo[];
  gateway: {
    version: string;
    uptime: number;
    activeConnections: number;
  };
}

// =============================================================================
// API Gateway Service
// =============================================================================

export class APIGatewayService {
  constructor(private readonly client: ACGS2Client) {}

  // ===========================================================================
  // Health & Status
  // ===========================================================================

  /**
   * Check API gateway health
   */
  async healthCheck(): Promise<HealthCheckResponse> {
    const response = await this.client.get<HealthCheckResponse>('/health');

    if (!response.success || !response.data) {
      throw new ValidationError('Health check failed');
    }

    return response.data;
  }

  // ===========================================================================
  // Feedback
  // ===========================================================================

  /**
   * Submit user feedback
   */
  async submitFeedback(request: FeedbackRequest): Promise<FeedbackResponse> {
    const response = await this.client.post<FeedbackResponse>('/feedback', {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to submit feedback');
    }

    return response.data;
  }

  /**
   * Get feedback statistics
   */
  async getFeedbackStats(): Promise<FeedbackStats> {
    const response = await this.client.get<FeedbackStats>('/feedback/stats');

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to get feedback stats');
    }

    return response.data;
  }

  // ===========================================================================
  // Service Discovery
  // ===========================================================================

  /**
   * List available services
   */
  async listServices(): Promise<ServicesResponse> {
    const response = await this.client.get<ServicesResponse>('/services');

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to list services');
    }

    return response.data;
  }
}
